### PR TITLE
[wip] Support and operator and grouping of logic/strategies

### PR DESCRIPTION
**Pull request by sveisvei**
Work-in-progress to support `groups` and `AND` operator.

# Groups

(wip)

# Operator support

(wip)

# Caveats
* no precedence between AND/OR, order precedence only.

Format notes/spec:
```
a: (1 AND 2 AND 3) OR 4

    group: [
        {
            group: [{1, operator: 'AND'}, {2, operator: 'AND'}, {3}]
        },
        {
            4,
        }
    ]

b: (1 AND 2) OR 3 OR 4

    group: [
        {
            group: [{1, operator: 'AND'}, {2}]
        },
        {
            3,
        }
        {
            4,
        }
    ]


c: 1 OR 2 OR 3 OR 4

        [
            {1}, {2}, {3}, {4}
        ]


d: (1 OR 2) AND 3 AND 4

    group: [
        {
            group: [
                {1},
                {2}
            ],
            operator: 'AND'
        },
        {
            3,
            operator: 'AND'
        }
        {
            4,
        }
    ]

e: 1 OR (2 AND 3 AND 4)

    [
        {1},
        {
            group: [
                {2, operator: 'AND'},
                {3, operator: 'AND'},
                {4}
            ]
        }
    ]
```

### CODE REVIEW

File: test/client.test.js [hunk 1]
```diff
@@ -196,3 +196,244 @@ test('should emit error when invalid feature runtime', t => {
...
+
+    t.true(result3 === false);
+});
+
+test('groups with inner AND group that resovles to true', t => {
+    const result = testGroup([
+        {
+            name: 'inner-group',
+            group: [{ name: 'custom', operator: 'AND' }, { name: 'custom' }],
+            operator: 'AND',
```
Comment (ivarconr) [inline test/client.test.js]: Does it ever make sense to have `operator` defined on the group-level if the code only looks on it on strategy-definition level inside groups?

File: src/strategy/strategy.ts [hunk 2]
```diff
@@ -17,4 +17,6 @@ export class Strategy {
 export interface StrategyTransportInterface {
     name: string;
     parameters: any;
+    operator?: string;
+    group?: StrategyTransportInterface[];
```
Comment (ivarconr) [inline src/strategy/strategy.ts]: Would it make sense to define a group-interface that extends the `StrategyTransportInterface` instead?

```typescript
export interface StrategyGroupTransportInterface extends StrategyTransportInterface {
    operator: string;
    group: StrategyTransportInterface[];
}
```

File: src/strategy/strategy.ts [hunk 3]
```diff
@@ -17,4 +17,6 @@ export class Strategy {
 export interface StrategyTransportInterface {
     name: string;
     parameters: any;
+    operator?: string;
```
Comment (ivarconr) [inline src/strategy/strategy.ts]: Could the operator be an `Enum` type instead? 

```typescript
enum StrategyGroupOperator {
    AND,
    OR
}
```

**Comment by ivarconr** [comment]
I think this looks promsing! 

I think there are a few typos in you spec, a), b) and d) the "outer" Array will never be a group? 

These should probably be:


**a: (1 AND 2 AND 3) OR 4**

```javascript
    [
        {
            group: [{1, operator: 'AND'}, {2, operator: 'AND'}, {3}]
        },
        {
            4,
        }
    ]
```

**b: (1 AND 2) OR 3 OR 4**

```javascript
    [
        {
            group: [{1, operator: 'AND'}, {2, operator: 'AND'}]
        },
        {
            3,
        }
        {
            4,
        }
    ]
```

**d: (1 OR 2) AND 3 AND 4**
```javascript
    [
        {
            group: [
                {1},
                {2}
            ],
            operator: 'AND'
        },
        {
            3,
            operator: 'AND'
        }
        {
            4,
        }
    ]
```

We should probably also require groups to have a `name` and `parameters` property. This will allow us to add the grouping capability, without breaking old clients, they will simply ignore groups. Remember that unleash promises that the unleash-api should always be one version backward compatible for clients, making it possible to upgrade unleash gradually.

File: src/strategy/strategy.ts (hunk 2 above)
Comment (ivarconr) [reply src/strategy/strategy.ts]: After revisiting the spec I realize that the operator should be on the `StrategyTransportInterface` and needs to be optional.

File: test/client.test.js (hunk 1 above)
Comment (sveisvei) [reply test/client.test.js]: I think we initially should target to not use operator on the top-level, but further down the line I dont see any point on supporting legacy.

File: src/strategy/strategy.ts (hunk 2 above)
Comment (sveisvei) [reply src/strategy/strategy.ts]: 👍

File: src/strategy/strategy.ts (hunk 3 above)
Comment (sveisvei) [reply src/strategy/strategy.ts]: Indeed!

File: test/client.test.js (hunk 1 above)
Comment (ivarconr) [reply test/client.test.js]: agreed :+1:

**Pull request closed without merging**